### DISCUSSION TITLE
PAT Basic 1015. 德才论 (25) (C语言实现) | PAT Solution in C

### DISCUSSION THREAD

**Discussion by xlucn** [xlucn/PAT#76] [General]
https://oliverlew.github.io/PAT/Basic/1015.html 

 题目

**Comment by 603970634** [comment]
老哥， 我看了你的代码 ，自己写了一遍， 为什么提交上去有三个测试点过不去，求指教。下面是代码
#include <stdio.h>
#include <stdlib.h>


    typedef struct student
    {
        int ID;
        int D;
        int C;
        int S;
        int rank;
    }STU;
    int rank(STU s,int L,int H)
    {
        if(s.D<L || s.C<L) return 0;
        else if(s.D>= H &&s.C>=H) return 1;
        else if(s.D>=H) return 2;
        else if(s.D>=s.C) return 3;
        else return 4;
    }
    int com(const void *a,const void *b)
    {
        STU s1=*(STU *)a;
        STU s2=*(STU *)b;
        if(s1.rank!=s2.rank)
            return s1.rank-s2.rank;
        else if(s1.S!=s2.S)
            return s2.S-s1.S;
        else if(s1.D!=s2.D)
            return s2.D-s1.D;
        else return s1.ID-s2.ID;
    }
    int main()
    {
        STU stu[100001];
        STU buffer[100001];
        int n,l,h;
        scanf("%d%d%d",&n,&l,&h);
        int i=0,k=0;
        for(;i<n;i++)
        {
            STU M=buffer[i];
            scanf("%d%d%d",&M.ID,&M.D,&M.C);
            M.S=M.D+M.C;
            if((M.rank=rank(M,l,h))!=0)
                stu[k++]=M;
        }
        printf("%d\n",k);
        qsort(stu,k,sizeof(STU),com);
        for(i=0;i<k;i++)
        {
            printf("%d %d %d\n",stu[i].ID,stu[i].D,stu[i].C);
        }

    }

**Comment by xlucn** [comment]
> @603970634

没仔细看，但我发现我们的rank函数不同，1-4分数顺序正好相反（注意一下哪个才是最高的），可能影响到排序结果不一样。

**Comment by 603970634** [comment]
感谢回复 ，不过已经找到BUG了，粗心了rank里面写成了具体数值。


------------------ 原始邮件 ------------------
发件人: "LuXu"<notifications@github.com>;
发送时间: 2019年7月4日(星期四) 晚上11:50
收件人: "OliverLew/oliverlew.github.io"<oliverlew.github.io@noreply.github.com>;
抄送: "天之☆骄子"<603970634@qq.com>;"Mention"<mention@noreply.github.com>;
主题: Re: [OliverLew/oliverlew.github.io] PAT Basic 1015. 德才论 (C语言实现) | PAT Solution in C (#17)



 
@603970634
  
没仔细看，但我发现我们的rank函数不同，1-4分数顺序正好相反（注意一下哪个才是最高的），可能影响到排序结果不一样。

**Comment by 603970634** [comment]
老哥，我看了你PAT1044那道题的代码有2个疑问，你有空能帮我看看吗。问题在代码的30行和54行，十分感谢。
#include <ctype.h>
#include <stdio.h>
#include <string.h>


char units[][20] = {"tret", "jan", "feb", "mar", "apr", "may", "jun", "jly",
    "aug", "sep", "oct", "nov", "dec"};
char tens[][20] = {"tam", "hel", "maa", "huh", "tou", "kes", "hei", "elo",
    "syy", "lok", "mer", "jou"};


int Mars2Earth(char *s)
{


    if(s)
    {
        int i;
        for( i = 0; i < 13; i++)         /* units digits */
            if(strcmp(s, units[i]) == 0)
                return i;
        for(i = 1; i < 13; i++)         /* tens digits */
            if(strcmp(s, tens[i - 1]) == 0)
                return i * 13;
    }
    return 0;
}


int main()
{
    int N,i, m;
    char line[11];
/*这里的代码是不是为了获取N的值并且吸收回车？为什么不能替换成下面这部分，替换之后提交进去答案全错
    gets(line);
    sscanf(line, "%d", &N);
*/
    scanf("%d",&N);
    fflush(stdin);
    
    for( i = 0; i < N; i++)
    {
        fgets(line, 11, stdin);
        if(isdigit(line[0]))                  /* Earth number */
        {
            sscanf(line, "%d", &m);
            if(m / 13 && m % 13)
                printf("%s %s\n", tens[m / 13 - 1], units[m % 13]);
            if(m / 13 && m % 13 == 0)
                printf("%s\n", tens[m / 13 - 1]);
            if(m / 13 == 0)
                printf("%s\n", units[m]);
        }
        if(isalpha(line[0]))                  /* Mars number */
        {
            m = Mars2Earth(strtok(line, " "));        /* higher digit */
            m += Mars2Earth(strtok(NULL, " "));       /* lower digit */
            //为什么要在空格符号加\n,把\n去掉之后在odeblocks和dev上输入elv nov的运行结果都是104（答案是115）,但是提交进PAT又全通
            //过了，十分疑惑，求大佬指点一下。
            printf("%d\n", m);
        }
    }


    return 0;
}





------------------ 原始邮件 ------------------
发件人: "LuXu"<notifications@github.com>;
发送时间: 2019年7月4日(星期四) 晚上11:50
收件人: "OliverLew/oliverlew.github.io"<oliverlew.github.io@noreply.github.com>;
抄送: "天之☆骄子"<603970634@qq.com>;"Mention"<mention@noreply.github.com>;
主题: Re: [OliverLew/oliverlew.github.io] PAT Basic 1015. 德才论 (C语言实现) | PAT Solution in C (#17)



 
@603970634
  
没仔细看，但我发现我们的rank函数不同，1-4分数顺序正好相反（注意一下哪个才是最高的），可能影响到排序结果不一样。

**Comment by xlucn** [comment]
@603970634

> ```c
> /*这里的代码是不是为了获取N的值并且吸收回车？为什么不能替换成下面这部分，替换之后提交进去答案全错
>     gets(line);
>     sscanf(line, "%d", &N);
> */
>     scanf("%d",&N);
>     fflush(stdin);
> ```
是。我目测应该是因为`scanf`不读取`\n`，导致后面`fgets`出错（它会读`\n`），并且flush输入很罕见，Google了一下，这篇文章说后果应该是未定义的，这不是处理换行符的常见方法。

https://www.geeksforgeeks.org/use-fflushstdin-c/

> ```c
>             m = Mars2Earth(strtok(line, " "));        /* higher digit */
>             m += Mars2Earth(strtok(NULL, " "));       /* lower digit */
>             //为什么要在空格符号加\n,把\n去掉之后在odeblocks和dev上输入elv nov的运行结果都是104（答案是115）,但是提交进PAT又全通
>             //过了，十分疑惑，求大佬指点一下。
> ```
这个应该也是读取到的字符串含`\n`的关系，这样截取到的单个数字就不会有。`strtok`的分隔符传入多个字符时的用法，你可以了解一下。

**Comment by 603970634** [comment]
谢了哥，明白了。后面那个错误是因为我之前把32行的fgets改成了gets,gets不会读取\n，所以分割符不用加\n也能通过。而你之前用的是fgets,读取了\n，所以分割符必须加\n，否则分割出来会多一个回车字符。




------------------ 原始邮件 ------------------
发件人: "LuXu"<notifications@github.com>;
发送时间: 2019年7月5日(星期五) 下午4:45
收件人: "OliverLew/oliverlew.github.io"<oliverlew.github.io@noreply.github.com>;
抄送: "天之☆骄子"<603970634@qq.com>;"Mention"<mention@noreply.github.com>;
主题: Re: [OliverLew/oliverlew.github.io] PAT Basic 1015. 德才论 (C语言实现) | PAT Solution in C (#17)




@603970634
  /*这里的代码是不是为了获取N的值并且吸收回车？为什么不能替换成下面这部分，替换之后提交进去答案全错     gets(line);     sscanf(line, "%d", &N); */     scanf("%d",&N);     fflush(stdin);
  
是。我目测应该是因为scanf不读取\n，导致后面fgets出错（它会读\n），并且flush输入很罕见，Google了一下，这篇文章说后果应该是未定义的，这不是处理换行符的常见方法。
 
https://www.geeksforgeeks.org/use-fflushstdin-c/
              m = Mars2Earth(strtok(line, " "));        /* higher digit */             m += Mars2Earth(strtok(NULL, " "));       /* lower digit */             //为什么要在空格符号加\n,把\n去掉之后在odeblocks和dev上输入elv nov的运行结果都是104（答案是115）,但是提交进PAT又全通             //过了，十分疑惑，求大佬指点一下。
  
这个应该也是读取到的字符串含\n的关系，这样截取到的单个数字就不会有。strtok的分隔符传入多个字符时的用法，你可以了解一下。
